### ISSUE TITLE
Task Groupings

### ISSUE THREAD

**Issue by cdcarter** [SFDO-Tooling/CumulusCI#772]
from @davisagli:
> it'd be nice if we had a way to group tasks in the list, and then maybe a way to hide a group unless you have a flag in your personal settings. In other words I want a way to say "show me all the releng tasks" rather than needing to configure each one

**Comment by cdcarter** [comment]
Humble proposal:

a task definition looks like:
```
tasks:
    blah_task_name:
        class_path: blah
        description: blah
        options:
            name: option_is_blah
```
[JSON Schema definition for the tasks section of cumulusci.yml](https://gist.github.com/cdcarter/d3a10e03f12265f1f49af2aa3d9c8c40#file-schema-json-L102-L129)

I suggest we add a fourth, optional, key to the task definition object:

```
tasks:
    blah_task_name:
        group: releng
        class_path: blah
        description: blah
        options:
            name: option_is_blah
```
defined as a comma seperated list of strings .

If a task defines a group (or more than one), by default, it will *not* be returned by BaseTaskFlowConfig.list_tasks.

We'll also add a new property to the cumulusci section, which contains a comma severated list of groups:
```
cumulusci:
    keychain: MyKeyChainClass
    groups: releng, wlma
```

If a group is added to the cumulusci.groups key, then BaseTaskFlowConfig.list_tasks will now list tasks from that group. This can be overriden on the project yaml, or on an (e.g.) release engineers loal machine.

Still to figure out:
- are groups pre-registered besides just being referenced in a task?
- the behavior of defining a task with the same name as a globally hidden task in a project and not specifying group would be surprising. perhaps all tasks must have a group.
- at the CLI you should be able to list *all* tasks, and list all tasks for a specified list of groups. this is not the default. there's also a way to force running a task in a group, if its not your default group. oh, and see them.
- but unless you've specified a group in YAML or at runtime, the task is invisible.
- docs should list all tasks all groups.

**Comment by davisagli** [comment]
Hmm, I'm having second thoughts about the idea of hiding groups of tasks because:
a. It'll be a source of confusion
b. It sort of goes against our ethos of making our internal tools widely available whenever they aren't truly SFDO Product Delivery-specific.
c. When there are tasks that fit the "SFDO Product Delivery-specific" clause, I'm inclined to say they should go in a separate package and we should have a way for installed Python packages to register tasks that they provide via setuptools entry points.

**Comment by cdcarter** [comment]
well, ignoring my "wlma" example, I think (with tweaks) it could still work out? 

reporting on push_failures isn't product delivery specific, it's just...
```
(cci-q4) ➜  npsp git:(master) ✗ cci task list | wc -l
      77
```
but we tell people to use the list/info tasks to get discoverability into what CCI can do. 

what if I walk the proposal back a little bit more...

If a task has a task group, it's not (by default) displayed on `cci task list`, unless that group is in your personal yaml. If it's in your yaml, it's displayed in the list. The list also prints at the bottom a line for each group and how many tasks it has that weren't displayed. `cci task list <group_name>` would do the obvious, as would `cci task list --all`

**Comment by davisagli** [comment]
I'm not keen on hiding-by-default any task that has a group, because I want to use the group to actually show the default results in groups, much like Django's manage.py help. I can see `cci task list <group_name>` being useful for focusing, but I'd kinda like to see how much grouping the results helps before going as far as to hide some groups by default.

**Comment by jstvz** [comment]
Right, maybe builtins could be categorized by topic (CI, Github, Deployment, etc), with another group (or column) to indicate project-specific customizations. It'd be a solid start.

**Comment by cdcarter** [comment]
Where I'm coming from is that unless you're the person doing packaging -- you don't need to see the five `push_*` tasks in the task list, they're just bloat. Things like `github_*` aren't really meant to be run by developers or even outside of CI. In fact, you mention django manage.py, which I feel has WAAAY too many lines of output, not enough information, and a frequently get annoyed at how its formatted :)

the proposal above has MOST tasks not in a group. that doesn't seem to be well loved...

What if every task has a group, and we have most groups enabled by default? But the few related to packaging orgs or CI only tasks are keyed out, and the task list does indicate there are more hidden tasks?

**Comment by davisagli** [comment]
> What if every task has a group, and we have most groups enabled by default? But the few related to packaging orgs or CI only tasks are keyed out, and the task list does indicate there are more hidden tasks?

This sounds pretty good to me.

**Comment by jstvz** [comment]
Closed by #872
